src/core/public/application/application_service.tsx

```diff
@@ -17,159 +17,157 @@
+import { HttpSetup } from '../http';
 import { ContextSetup, IContextContainer } from '../context';
+import { AppRouter } from './ui';
+import { CapabilitiesService, Capabilities } from './capabilities';
 import {
   App,
   LegacyApp,
   AppMounter,
+  LegacyAppMounter,
+  Mounter,
   InternalApplicationSetup,
   InternalApplicationStart,
 } from './types';
 
 interface SetupDeps {
   context: ContextSetup;
-}
-
-interface StartDeps {
-  http: HttpStart;
-  injectedMetadata: InjectedMetadataStart;
+  http: HttpSetup;
+  injectedMetadata: InjectedMetadataSetup;
   /**
    * Only necessary for redirecting to legacy apps
    * @deprecated
    */
   redirectTo?: (path: string) => void;
 }
 
-interface AppBox {
-  app: App;
-  mount: AppMounter;
+interface StartDeps {
+  injectedMetadata: InjectedMetadataStart;
 }
 
+const filterAvailable = (map: Map<string, any>, capabilities: Capabilities) =>
+  new Map(
+    [...map].filter(
+      ([id]) => capabilities.navLinks[id] === undefined || capabilities.navLinks[id] === true
+    )
+  );
+const hasAppBasePath = (mounters: Map<string, Mounter>, appBasePath?: string) =>
+  [...mounters.values()].some(mounter => mounter.appBasePath === appBasePath);
+const getAppUrl = (mounters: Map<string, Mounter>, appId: string, path: string = '') =>
+  `/${mounters.get(appId)?.appBasePath ?? `/app/${appId}`}/${path}`
+    .replace(/\/{2,}/g, '/') // Remove duplicate slashes
+    .replace(/\/$/, ''); // Remove trailing slash
+
 /**
  * Service that is responsible for registering new applications.
  * @internal
  */
 export class ApplicationService {
-  private readonly apps$ = new BehaviorSubject<ReadonlyMap<string, AppBox>>(new Map());
-  private readonly legacyApps$ = new BehaviorSubject<ReadonlyMap<string, LegacyApp>>(new Map());
+  private readonly apps = new Map<string, App>();
+  private readonly legacyApps = new Map<string, LegacyApp>();
+  private readonly mounters = new Map<string, Mounter>();
   private readonly capabilities = new CapabilitiesService();
+  private currentAppId$ = new BehaviorSubject<string | undefined>(undefined);
+  private registrationClosed = false;
+  private history: History<any> | null = null;
   private mountContext?: IContextContainer<App['mount']>;
+  private navigate?: (url: string, state: any) => void;
+
+  public setup({
+    context,
+    http: { basePath },
+    injectedMetadata,
+    redirectTo = (path: string) => (window.location.href = path),
+  }: SetupDeps): InternalApplicationSetup {
+    // Only setup history if we're not in legacy mode
+    if (!injectedMetadata.getLegacyMode()) {
+      this.history = createBrowserHistory({ basename: basePath.get() });
+    }
 
-  public setup({ context }: SetupDeps): InternalApplicationSetup {
+    // If we do not have history available, use redirectTo to do a full page refresh.
+    this.navigate = (url, state) =>
+      this.history ? this.history.push(url, state) : redirectTo(basePath.prepend(url));
```

Comment:
we don't need to `basePath.prepend` for `history.push(url)`?

Reply:
Correct, I'll add back the comment that existed in the previous implementation.